sklearn/preprocessing/tests/test_discretization.py

```diff
@@ -281,3 +281,22 @@ def test_percentile_numeric_stability():
     assert_warns_message(UserWarning, msg, kbd.fit, X)
     assert_array_almost_equal(kbd.bin_edges_[0], bin_edges)
     assert_array_almost_equal(kbd.transform(X), Xt)
+
+
+@pytest.mark.parametrize("input_dtype", [np.float16, np.float32, np.float64])
+@pytest.mark.parametrize("output_dtype", [np.float16, np.float32, np.float64])
```

Comment:
This is Cartesian product between input and output dtype. What we want is,
```py
@pytest.mark.parametrize("input_dtype, output_dtype", 
                         zip([np.float16, np.float32, np.float64])
                             [np.float16, np.float32, np.float64])))
```

I don't understand why it doesn't fail in its current form. It should I think.

Reply:
It does not fail because I decorrelate the output dtype from the input dtype. In my first solution, I want the output dtype to match the dtype defined at the class level (in the `__init__`). So whatever my input dtype is, I want my output dtype to match the one I define in when I call `KBinsDiscretizer`. We have the same behaviour in `OneHotEncoder.transform()`.  but if you prefer to make the dtype consistent with the input of `transform`, it is possible ! In that case a new PR for `OneHotEncoder` might be considered as well.

Reply:
Indeed you are right, never mind my comment. The current solution is correct.